### PR TITLE
Address CVE-2022-1996

### LINKED ISSUE #369: Version 3.5.0 vulnerability with CVE-2022-1996
Hello, release 3.5.0 is apparently vulnerable with a score of 9+ https://nvd.nist.gov/vuln/detail/CVE-2022-1996

NIST CVSS scoreNIST: NVDBase Score: [9.1 CRITICAL](https://nvd.nist.gov/vuln-metrics/cvss/v3-calculator?name=CVE-2022-1996&vector=AV:N/AC:L/PR:N/UI:N/S:U/C:H/I:H/A:N&version=3.1&source=NIST)Vector:  CVSS:3.1/AV:N/AC:L/PR:N/UI:N/S:U/C:H/I:H/A:N

Nist CVSS score does not match with CNA score
CNA:  huntr.devBase Score: [9.3 CRITICAL](https://nvd.nist.gov/vuln-metrics/cvss/v3-calculator?name=CVE-2022-1996&vector=AV:N/AC:L/PR:N/UI:R/S:C/C:H/I:H/A:N&version=3.0&source=huntr.dev)

### PR DESCRIPTION
**Pull request by datamattsson**
Signed-off-by: Michael Mattsson <michael.mattsson@gmail.com>

**What type of PR is this?**
> Uncomment only one ` /kind <>` line, hit enter to put that in a new line, and remove leading whitespaces from that line:
>
> /kind api-change
> /kind bug
/kind cleanup
> /kind design
> /kind documentation
> /kind failing-test
> /kind feature
> /kind flake

**What this PR does / why we need it**:

Addresses CVE-2022-1996

**Which issue(s) this PR fixes**:

Fixes #369

**Special notes for your reviewer**:

Slack thread: https://kubernetes.slack.com/archives/C09QZFCE5/p1670445226176689

**Does this PR introduce a user-facing change?**:

```release-note
NONE
```

### CODE REVIEW

**Comment by bswartz** [comment]
Is it possible to not bump the go version from 1.16 to 1.18? For a stable release we don't want to change unnecessary things like compiler versions, due to the chance that the new compiler breaks something that used to work (due to deprecation).

File: go.mod [hunk 1]
```diff
@@ -1,77 +1,67 @@
...
-	gopkg.in/yaml.v3 v3.0.1 // indirect
-	k8s.io/api v0.24.0
-	k8s.io/apimachinery v0.24.0
-	k8s.io/client-go v0.24.0
-	k8s.io/component-base v0.23.0 // indirect
-	k8s.io/csi-translation-lib v0.23.6
-	k8s.io/klog/v2 v2.60.1
+	github.com/kubernetes-csi/csi-test/v4 v4.4.0
+	google.golang.org/grpc v1.51.0
+	k8s.io/api v0.26.0-rc.0
```
[the remark this answers is not in the document]
Comment (datamattsson) [reply go.mod]: > I think we want the kubernetes dependencies to stay in the same minor version (we can update to a newer patch version)

Done.

**Comment by datamattsson** [comment]
> Is it possible to not bump the go version from 1.16 to 1.18?

1.16 has a cascade of CVEs associated with it in some of the indirect dependencies. Is 1.17 a middle road?

**Comment by bswartz** [comment]
> > Is it possible to not bump the go version from 1.16 to 1.18?
> 
> 1.16 has a cascade of CVEs associated with it in some of the indirect dependencies. Is 1.17 a middle road?

Even the latest dot release - 1.16.15? If there are known unpatched security vulnerabilities then I agree it wouldn't make much sense to continue using the old compiler version in a patch that specifically addresses security vulnerabilities, but we should want to avoid upgrading anything that might deprecate existing support because nobody wants to do a dot release upgrade (of this sidecar) and find that something broke.

I'm not specifically aware of anything that was dropped between go version 1.16 and 1.17/1.18, but that's the thing to worry about.

**Comment by datamattsson** [comment]
> Even the latest dot release - 1.16.15?

Changing go.mod from 1.17 to 1.16 and running `go mod tidy && go mod vendor` in between will reveal this report: https://gist.github.com/datamattsson/e1038144465ace0177dd8b904e9de278

Yesterday, 1.17 was clean as a whistle but today [CVE-2022-41717](https://avd.aquasec.com/nvd/cve-2022-41717) in golang.org/x/net was discovered (not relevant to the attacher AFAICT).

**Comment by datamattsson** [comment]
> nobody wants to do a dot release upgrade (of this sidecar) and find that something broke.

This is why we test. Any sidecar upgrade, major, minor or patch would require full e2e on our end and expect the same results. I don't think users blindly flick an image upgrade before safely testing the new image.

**Comment by bswartz** [comment]
> > nobody wants to do a dot release upgrade (of this sidecar) and find that something broke.
> 
> This is why we test. Any sidecar upgrade, major, minor or patch would require full e2e on our end and expect the same results. I don't think users blindly flick an image upgrade before safely testing the new image.

Given the CVEs associated with older go compiler versions, I'm not going to oppose an update here, but did want to push back on the idea that testing is sufficient in every case. Testing covers a lot of use cases but typically only in a single (or small number of) configurations. More exotic configurations tend not to be tested, and deprecations tend to affect exotic configurations more than anything. It's a good rule in general to avoid updating things in stable branches without an explicit reason. In this case our explicit reason is security issues with the compiler, so I'm fine with that.

**Comment by bswartz** [comment]
> This is why we test. Any sidecar upgrade, major, minor or patch would require full e2e on our end and expect the same results. I don't think users blindly flick an image upgrade before safely testing the new image.

Given the CVEs associated with older go compiler versions, I'm not going to oppose an update here, but did want to push back on the idea that testing is sufficient in every case. Testing covers a lot of use cases but typically only in a single (or small number of) configurations. More exotic configurations tend not to be tested, and deprecations tend to affect exotic configurations more than anything. It's a good rule in general to avoid updating things in stable branches without an explicit reason. In this case our explicit reason is security issues with the compiler, so I'm fine with that.

**Comment by datamattsson** [comment]
Will this PR be merge-able at one point? (I understand 1.26 deliverables takes precedence).

**Comment by jsafrane** [comment]
I'm fixing CI in https://github.com/kubernetes-csi/external-attacher/pull/401

**Comment by jsafrane** [comment]
> Is it possible to not bump the go version from 1.16 to 1.18? 

version in go.mod just specifies version of go vendoring rules that will be used by `go mod`. The actual compiler used to build the image is defined either in [cloudbuild.yaml](https://github.com/kubernetes-csi/external-attacher/blob/master/release-tools/cloudbuild.yaml#L29) or in [prow.sh](https://github.com/kubernetes-csi/external-attacher/blob/0e00c8ada1ecbcd2557bd3d6445286001b7f7a4e/release-tools/prow.sh#L89), both taken from the latest release-tools merge. It's not bound to the branch in any way. I can't tell if it is good or bad, just reporting the process.

**Pull request merged**